File: EPD290_4C/RPI/python/lib/lafvin_epd/epd2in9g.py
"""Driver for the LAFVIN 2.9-inch, 128 x 296, four-color e-paper."""

from __future__ import annotations

import logging
import time

from PIL import Image

from . import epdconfig

logger = logging.getLogger(__name__.split(".")[-1])

EPD_WIDTH = 128
EPD_HEIGHT = 296
FRAME_BYTES = EPD_WIDTH * EPD_HEIGHT // 4

# Pillow uses RGB tuples. The controller later converts these colors to 2-bit codes.
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
YELLOW = (255, 255, 0)
RED = (255, 0, 0)

CMD_PANEL_SETTING = 0x00
CMD_POWER_ON = 0x04
CMD_POWER_OFF = 0x02
CMD_DATA_START = 0x10
CMD_DISPLAY_REFRESH = 0x12
CMD_DEEP_SLEEP = 0x07

BUSY_TIMEOUT_SECONDS = 120.0
BUSY_POLL_MS = 5

try:
    DITHER_FLOYDSTEINBERG = Image.Dither.FLOYDSTEINBERG
    DITHER_NONE = Image.Dither.NONE
    ROTATE_90 = Image.Transpose.ROTATE_270
    ROTATE_180 = Image.Transpose.ROTATE_180
    ROTATE_270 = Image.Transpose.ROTATE_90
except AttributeError:  # Pillow < 9.1
    DITHER_FLOYDSTEINBERG = Image.FLOYDSTEINBERG
    DITHER_NONE = Image.NONE
    ROTATE_90 = Image.ROTATE_270
    ROTATE_180 = Image.ROTATE_180
    ROTATE_270 = Image.ROTATE_90


class EPD:
    """High-level display driver used by the teaching example."""

    width = EPD_WIDTH
    height = EPD_HEIGHT
    BLACK = BLACK
    WHITE = WHITE
    YELLOW = YELLOW
    RED = RED

    def __init__(self, hardware=epdconfig):
        self._hardware = hardware
        self.reset_pin = hardware.RST_PIN
        self.dc_pin = hardware.DC_PIN
        self.busy_pin = hardware.BUSY_PIN

    def reset(self):
        """Generate the high-low-high reset pulse required by the controller."""
        self._hardware.digital_write(self.reset_pin, 1)
        self._hardware.delay_ms(200)
        self._hardware.digital_write(self.reset_pin, 0)
        self._hardware.delay_ms(2)
        self._hardware.digital_write(self.reset_pin, 1)
        self._hardware.delay_ms(200)

    def send_command(self, command):
        """Send one controller command (DC = 0)."""
        self._hardware.digital_write(self.dc_pin, 0)
        self._hardware.spi_write(bytes((command & 0xFF,)))

    def send_data(self, data):
        """Send one byte or a byte sequence as controller data (DC = 1)."""
        payload = bytes((data & 0xFF,)) if isinstance(data, int) else bytes(data)
        self._hardware.digital_write(self.dc_pin, 1)
        self._hardware.spi_write(payload)

    def wait_until_ready(self, timeout=BUSY_TIMEOUT_SECONDS):
        """Wait until BUSY becomes high, or raise TimeoutError."""
        logger.debug("e-Paper busy H")
        self._hardware.delay_ms(100)
        deadline = time.monotonic() + timeout

        while self._hardware.digital_read(self.busy_pin) == 0:
            if time.monotonic() >= deadline:
                raise TimeoutError(f"e-Paper BUSY timed out after {timeout:g} seconds")
            self._hardware.delay_ms(BUSY_POLL_MS)

        logger.debug("e-Paper busy H release")

    def _configure_panel(self):
        """Send the common initialization commands in datasheet order."""
        # Send each command before its associated data. Use the initialization
        # values specified for the exact panel model.
        self.send_command(0x4D)
        self.send_data(0x78)

        self.send_command(CMD_PANEL_SETTING)  # Panel setting
        self.send_data((0x0F, 0x29))

        self.send_command(0x01)  # Power setting
        self.send_data((0x07, 0x00))

        self.send_command(0x03)  # Power-off sequence
        self.send_data((0x10, 0x54, 0x44))

        self.send_command(0x06)  # Booster soft start
        self.send_data((0x0F, 0x0A, 0x2F, 0x25, 0x22, 0x2E, 0x21))

        self.send_command(0x41)  # Temperature sensor setting
        self.send_data(0x00)

        self.send_command(0x50)  # VCOM and data interval
        self.send_data(0x37)

        self.send_command(0x60)  # TCON setting
        self.send_data((0x02, 0x02))

        self.send_command(0x61)  # Resolution: 128 x 296
        self.send_data(
            (
                self.width >> 8,
                self.width & 0xFF,
                self.height >> 8,
                self.height & 0xFF,
            )
        )

        self.send_command(0x65)  # Gate/source start position
        self.send_data((0x00, 0x00, 0x00, 0x00))

        # Controller internal drive parameters; users normally should not change these.
        self.send_command(0xE7)
        self.send_data(0x1C)
        self.send_command(0xE3)
        self.send_data(0x22)
        self.send_command(0xB4)
        self.send_data(0xD0)
        self.send_command(0xB5)
        self.send_data(0x03)
        self.send_command(0xE9)
        self.send_data(0x01)
        self.send_command(0x30)
        self.send_data(0x08)

    def init(self):
        """Initialize the panel for a normal full refresh."""
        self._hardware.module_init()
        self.reset()
        self.wait_until_ready()
        self._configure_panel()

        self.send_command(CMD_POWER_ON)
        self._hardware.delay_ms(500)
        self.wait_until_ready()
        return 0

    def init_fast(self):
        """Initialize the panel for the controller's fast refresh mode."""
        self._hardware.module_init()
        self.reset()
        self.wait_until_ready()
        self._configure_panel()

        # Fast-refresh-specific settings.
        self.send_command(0xE0)
        self.send_data(0x02)
        self.send_command(0xE6)
        self.send_data(90)
        self.send_command(0xA5)
        self.wait_until_ready()

        self.send_command(CMD_POWER_ON)
        self.wait_until_ready()
        return 0

    def _turn_on_display(self):
        """Apply the image already stored in controller RAM to the panel."""
        self.send_command(CMD_DISPLAY_REFRESH)
        self.send_data(0x00)
        self.wait_until_ready()

    @staticmethod
    def _palette_image():
        palette = Image.new("P", (1, 1))
        colors = [
            0, 0, 0,        # 0: black
            255, 255, 255,  # 1: white
            255, 255, 0,    # 2: yellow
            255, 0, 0,      # 3: red
        ]
        colors.extend([0, 0, 0] * 252)
        palette.putpalette(colors)
        return palette

    @staticmethod
    def _flatten_to_white(image):
        if image.mode in ("RGBA", "LA") or "transparency" in image.info:
            rgba = image.convert("RGBA")
            flattened = Image.new("RGB", rgba.size, WHITE)
            flattened.paste(rgba, mask=rgba.getchannel("A"))
            return flattened
        return image.convert("RGB")

    def getbuffer(self, image, rotation=0, dither=False):
        """Convert a Pillow image into a native 9,472-byte frame."""
        rotations = {0: None, 90: ROTATE_90, 180: ROTATE_180, 270: ROTATE_270}
        if rotation not in rotations:
            raise ValueError("rotation must be 0, 90, 180 or 270 degrees")

        expected_size = (
            (self.width, self.height)
            if rotation in (0, 180)
            else (self.height, self.width)
        )
        if image.size != expected_size:
            raise ValueError(
                f"invalid image size {image.width}x{image.height}; expected "
                f"{expected_size[0]}x{expected_size[1]} for rotation {rotation}"
            )

        image_temp = self._flatten_to_white(image)
        if rotations[rotation] is not None:
            image_temp = image_temp.transpose(rotations[rotation])

        dither_mode = DITHER_FLOYDSTEINBERG if dither else DITHER_NONE
        image_4color = image_temp.quantize(
            palette=self._palette_image(), dither=dither_mode
        )
        color_codes = bytearray(image_4color.tobytes())

        if len(color_codes) != self.width * self.height:
            raise ValueError("Pillow returned an unexpected pixel count")

        frame = bytearray(FRAME_BYTES)
        for source in range(0, len(color_codes), 4):
            # Palette entries 4..255 are duplicate black entries.
            pixels = [
                value if value < 4 else 0
                for value in color_codes[source : source + 4]
            ]
            frame[source // 4] = (
                (pixels[0] << 6)
                | (pixels[1] << 4)
                | (pixels[2] << 2)
                | pixels[3]
            )
        return frame

    def display(self, image, rotation=0, dither=False):
        """Display a Pillow image or an already-packed native frame."""
        if isinstance(image, Image.Image):
            frame = self.getbuffer(image, rotation=rotation, dither=dither)
        else:
            frame = bytes(image)
            if rotation:
                raise ValueError("rotation is only supported for Pillow images")
        if len(frame) != FRAME_BYTES:
            raise ValueError(
                f"display buffer must contain exactly {FRAME_BYTES} bytes"
            )

        self.send_command(CMD_DATA_START)
        self.send_data(frame)
        self._turn_on_display()

    @staticmethod
    def _packed_clear_color(color):
        rgb_to_code = {BLACK: 0, WHITE: 1, YELLOW: 2, RED: 3}
        if isinstance(color, (tuple, list)):
            try:
                code = rgb_to_code[tuple(color)]
            except KeyError as exc:
                raise ValueError("clear color must be BLACK, WHITE, YELLOW or RED") from exc
            return code * 0x55

        if isinstance(color, int):
            if 0 <= color <= 3:
                return color * 0x55
            if color in (0x55, 0xAA, 0xFF):
                return color
        raise ValueError(
            "clear color must be a four-color RGB value, code 0..3, "
            "or packed byte 0x55/0xAA/0xFF"
        )

    def clear(self, color=WHITE):
        """Fill the entire panel with one of the four supported colors."""
        packed = self._packed_clear_color(color)
        self.send_command(CMD_DATA_START)
        self.send_data(bytes((packed,)) * FRAME_BYTES)
        self._turn_on_display()

    def Clear(self, color=1):
        """Compatibility alias for scripts written against the legacy Clear() spelling."""
        self.clear(color)

    def sleep(self):
        """Power off the controller, enter deep sleep, and release GPIO/SPI."""
        try:
            self.send_command(CMD_POWER_OFF)
            self.send_data(0x00)
            self.wait_until_ready()

            self.send_command(CMD_DEEP_SLEEP)
            self.send_data(0xA5)
            self._hardware.delay_ms(2000)
        finally:
            self.close()

    def close(self):
        """Release hardware resources without sending panel sleep commands."""
        self._hardware.module_exit()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

File: EPD290_4C/RPI/c/lib/e-Paper/EPD_2in9g.c
#include "EPD_2in9g.h"

#include <stddef.h>
#include <string.h>

#include "Debug.h"

#define CMD_PANEL_SETTING 0x00U
#define CMD_POWER_ON 0x04U
#define CMD_DATA_START 0x10U
#define CMD_DISPLAY_REFRESH 0x12U
#define CMD_POWER_OFF 0x02U
#define CMD_DEEP_SLEEP 0x07U

#define BUSY_TIMEOUT_MS 120000U
#define BUSY_POLL_MS 5U

/*
 *
 * SPI carries both commands and data. DC selects the transfer type:
 * DC = 0 for a command byte and DC = 1 for a data byte.
 */
typedef struct {
    UBYTE command; /* Register command. */
    UBYTE length;  /* Number of data bytes that follow. */
    UBYTE data[7]; /* This panel uses at most seven data bytes per init command. */
} RegisterWrite;

/*
 * Common initialization table used by both full and fast refresh modes.
 *
 * For example, {0x4D, 1, {0x78}} sends command 0x4D followed by data 0x78.
 *
 * {0x06, 7, {...}} sends command 0x06 followed by seven data bytes.
 *
 * These values come from the controller initialization sequence. Do not copy
 * them to another panel without checking that panel's datasheet or driver.
 */
static const RegisterWrite common_init[] = {
    {0x4D, 1, {0x78}},                                  /* Internal setting. */
    {CMD_PANEL_SETTING, 2, {0x0F, 0x29}},               /* Panel mode. */
    {0x01, 2, {0x07, 0x00}},                            /* Power setting. */
    {0x03, 3, {0x10, 0x54, 0x44}},                      /* Power-off sequence. */
    {0x06, 7, {0x0F, 0x0A, 0x2F, 0x25, 0x22, 0x2E, 0x21}}, /* Booster soft start. */
    {0x41, 1, {0x00}},                                  /* Temperature sensor. */
    {0x50, 1, {0x37}},                                  /* VCOM/data interval. */
    {0x60, 2, {0x02, 0x02}},                            /* TCON setting. */
    {0x61, 4, {                                          /* Resolution: 128 x 296. */
        (UBYTE)(EPD_2IN9G_WIDTH >> 8),
        (UBYTE)(EPD_2IN9G_WIDTH & 0xFF),
        (UBYTE)(EPD_2IN9G_HEIGHT >> 8),
        (UBYTE)(EPD_2IN9G_HEIGHT & 0xFF)
    }},
    {0x65, 4, {0x00, 0x00, 0x00, 0x00}},                /* Gate/source start. */
    /* Internal controller drive parameters; users normally should not edit them. */
    {0xE7, 1, {0x1C}},
    {0xE3, 1, {0x22}},
    {0xB4, 1, {0xD0}},
    {0xB5, 1, {0x03}},
    {0xE9, 1, {0x01}},
    {0x30, 1, {0x08}},
};

static void reset_panel(void)
{
    /* Generate the high-low-high reset pulse required by the controller. */
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(200);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(2);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(200);
}

static int write_bytes(UBYTE data_mode, const UBYTE *data, size_t length)
{
    /* data_mode selects command (0) or data (1); hardware SPI controls CS. */
    DEV_Digital_Write(EPD_DC_PIN, data_mode);
    return DEV_SPI_Write(data, length);
}

static int write_command(UBYTE command)
{
    return write_bytes(0, &command, 1);
}

static int write_data(const UBYTE *data, size_t length)
{
    return write_bytes(1, data, length);
}

static int write_register(const RegisterWrite *entry)
{
    if (write_command(entry->command) < 0) {
        return -1;
    }
    return entry->length == 0 ? 0 : write_data(entry->data, entry->length);
}

static int wait_busy_level(UBYTE ready_level, const char *label)
{
    /*
     * BUSY reports whether the controller has completed its internal work.
     * This panel is ready when BUSY is high at these wait points.
     * The timeout prevents wiring faults from blocking the program forever.
     */
    Debug("e-Paper busy %s\n", label);
    DEV_Delay_ms(100);

    UDOUBLE elapsed = 100;
    while (DEV_Digital_Read(EPD_BUSY_PIN) != ready_level) {
        if (elapsed >= BUSY_TIMEOUT_MS) {
            Debug("e-Paper busy timeout (%s, %u ms)\n", label, BUSY_TIMEOUT_MS);
            return -1;
        }
        DEV_Delay_ms(BUSY_POLL_MS);
        elapsed += BUSY_POLL_MS;
    }

    Debug("e-Paper busy %s release\n", label);
    return 0;
}

static int write_common_init(void)
{
    /* Send entries in table order; register order must not be changed. */
    for (size_t i = 0; i < sizeof(common_init) / sizeof(common_init[0]); ++i) {
        if (write_register(&common_init[i]) < 0) {
            Debug("EPD register write failed: 0x%02X\n", common_init[i].command);
            return -1;
        }
    }
    return 0;
}

static int initialize_panel(int fast)
{
    /* Full and fast initialization share the panel configuration above. */
    reset_panel();
    if (wait_busy_level(1, "H") < 0 || write_common_init() < 0) {
        return -1;
    }

    if (fast) {
        /* Fast refresh requires additional waveform and timing settings. */
        const RegisterWrite fast_init[] = {
            {0xE0, 1, {0x02}},
            {0xE6, 1, {90}},
            {0xA5, 0, {0}},
        };
        for (size_t i = 0; i < sizeof(fast_init) / sizeof(fast_init[0]); ++i) {
            if (write_register(&fast_init[i]) < 0) {
                return -1;
            }
        }
        if (wait_busy_level(1, "H") < 0) {
            return -1;
        }
    }

    if (write_command(CMD_POWER_ON) < 0) {
        return -1;
    }
    if (!fast) {
        DEV_Delay_ms(500);
    }
    return wait_busy_level(1, "H");
}

static int refresh_panel(void)
{
    /* Refresh only after the image has been written to controller RAM. */
    const UBYTE value = 0x00;
    if (write_command(CMD_DISPLAY_REFRESH) < 0 || write_data(&value, 1) < 0) {
        return -1;
    }
    return wait_busy_level(1, "H");
}

void EPD_2IN9G_Init(void)
{
    if (initialize_panel(0) < 0) {
        Debug("EPD full initialization failed\n");
    }
}

void EPD_2IN9G_Init_Fast(void)
{
    if (initialize_panel(1) < 0) {
        Debug("EPD fast initialization failed\n");
    }
}

void EPD_2IN9G_Clear(UBYTE color)
{
    /*
     * Each four-color pixel uses 2 bits, so one byte stores four pixels.
     * A white color code of 1 packs as 01 01 01 01, or 0x55.
     */
    UBYTE packed = (UBYTE)((color & 0x03U) * 0x55U);
    UBYTE chunk[256];
    memset(chunk, packed, sizeof(chunk));

    if (write_command(CMD_DATA_START) < 0) {
        return;
    }

    size_t remaining = EPD_2IN9G_FRAME_BYTES;
    while (remaining != 0) {
        size_t length = remaining < sizeof(chunk) ? remaining : sizeof(chunk);
        if (write_data(chunk, length) < 0) {
            return;
        }
        remaining -= length;
    }
    (void)refresh_panel();
}

void EPD_2IN9G_Display(const UBYTE *image)
{
    if (image == NULL) {
        Debug("EPD display skipped: image is NULL\n");
        return;
    }
    /*
     * One frame is 128 x 296 x 2 bits = 9,472 bytes.
     * Write the complete frame before calling refresh_panel().
     */
    if (write_command(CMD_DATA_START) < 0 ||
        write_data(image, EPD_2IN9G_FRAME_BYTES) < 0) {
        return;
    }
    (void)refresh_panel();
}

void EPD_2IN9G_Sleep(void)
{
    /* Power off, wait for completion, and then enter deep sleep. */
    const UBYTE zero = 0x00;
    const UBYTE sleep_key = 0xA5;

    if (write_command(CMD_POWER_OFF) < 0 || write_data(&zero, 1) < 0) {
        return;
    }
    if (wait_busy_level(1, "H") < 0) {
        return;
    }
    if (write_command(CMD_DEEP_SLEEP) < 0 || write_data(&sleep_key, 1) < 0) {
        Debug("EPD deep sleep command failed\n");
    }
}
